# Alpaca Crypto Puzzle Solve
### Solved by Paden Roder on 8/31/2021
---
## 1) Alpaca Crypto Puzzle Introduction
I was emailed by Alpaca informing me that I can join the waitlist of their Crytpo Beta waitlist! If interested in joining the waitlist, use [this link!](https://alpaca.markets/crypto-invite?r=fja5Q)

Scrolling down on the invite page, I saw a section for their coding challenge:
<br><img src="screenshots/AlpacaChallenge.png" alt="Alpaca Crypto Challenge Prompt" title="Alpaca Crypto Challenge Prompt" /><br>

Step 2 suggests that it wants me to execute a GET request to the endpoint URL = (```https://api.alpaca.markets/v2/puzzle/202108/hint/{s}```), where ```{s}``` is given by the "decoded" string ```'MzEzMTMwMmQ2NDYxNzk3MzJkNjE2Njc0NjU3MjJkNmM2NTY4NmQ2MTZlMmQ2MzZmNmM2YzYxNzA3MzY1NjQyZTZhNzA2Nwo='```. The question followed: how to decode ```{s}```.

## 2) Decode the String
Honestly, this step would have been a lot more difficult for someone who hasn't seen base64 encoded data before, which this looks like. To test this is the case, I try below to:
1. Encode the base64 string into a bytes of ASCII encoding.
2. Then, we decode the ASCII bytes from base64 to our original message type.
2. Last, we decode the message bytes into a human-readable string.

In [126]:
import base64

b64encoded_string = 'MzEzMTMwMmQ2NDYxNzk3MzJkNjE2Njc0NjU3MjJkNmM2NTY4NmQ2MTZlMmQ2MzZmNmM2YzYxNzA3MzY1NjQyZTZhNzA2Nwo='
b64_bytes = b64encoded_string.encode('ascii')
msg_bytes = base64.b64decode(b64_bytes)
msg_string = msg_bytes.decode('ascii')
print('Decoded string: ' + str(msg_string))

Decoded string: 3131302d646179732d61667465722d6c65686d616e2d636f6c6c61707365642e6a7067



Hmm... Well, ```3131302d646179732d61667465722d6c65686d616e2d636f6c6c61707365642e6a7067``` I guess *could* be the decoded string... but I'm doubtful. Let's give it a try in a GET request anyways, to be sure...

## 3) Try GET Request to https://api.alpaca.markets/v2/puzzle/202108/hint/3131302d646179732d61667465722d6c65686d616e2d636f6c6c61707365642e6a7067
Let's try this GET request:

In [127]:
import requests
import config #* Contains my personal keys, you'll want to create a config.py file with your own Alpaca keys stored and put in the same directory as this IPYNB file if you'd like to follow along.

get_request_url = 'https://paper-api.alpaca.markets/v2/puzzle/202108/hint/' + str(msg_string)
print("GET request URL: " + get_request_url)

#* Perfrom GET request on this URL:
headers = {'APCA-API-KEY-ID': config.API_KEY, 'APCA-API-SECRET-KEY': config.API_SECRET}

r = requests.get(get_request_url, headers=headers)
print("Request JSON response: " + str(r.json()))

GET request URL: https://paper-api.alpaca.markets/v2/puzzle/202108/hint/3131302d646179732d61667465722d6c65686d616e2d636f6c6c61707365642e6a7067

Request JSON response: {'code': 40010001, 'message': 'given image does not exist - try again.'}


## 4) What Went Wrong? HINT: Finish Decoding the String!
As you can see, the response we got above was ```{'code': 40010001, 'message': 'given image does not exist - try again.'}```. So, it seems like we decoded our initial string ```s``` incorrectly.

After thinking on this and looking at the output, it seems like our decoded ```msg_string``` may not be *fully* decoded. Specifically, when I look at the decoded string ```3131302d646179732d61667465722d6c65686d616e2d636f6c6c61707365642e6a7067```, it looks like it is in hexademical (alpha-numeric, but with no characters > f). 

Let's see if ```msg_string``` can be further decoded from hexidemical to ASCII:

In [128]:
final_msg_string = bytearray.fromhex(msg_string).decode()
print('msg_string decoded from hex to ASCII: ' + final_msg_string)

msg_string decoded from hex to ASCII: 110-days-after-lehman-collapsed.jpg


Eureka! Now we're cooking!

## 5) Understanding the Output
OK, I'm certain we have the answer to the decoded string now. ```110-days-after-lehman-collapsed.jpg``` would make sense to be at the end of a GET request. However, what does the filename mean?

The [Lehman Brothers](https://en.wikipedia.org/wiki/Lehman_Brothers) filed for bankruptcy on September 15, 2008; signalling the start of the Great Recession. 110 days after this date is January 3, 2009. One this date, **the Bitcoin network is created as the first block of the digital currency is mined by a person or group of people using the name Satoshi Nakamoto.** 

Yeah, we're on the right track here. Let's try the GET request.

## 6) Try GET Request to https://api.alpaca.markets/v2/puzzle/202108/hint/110-days-after-lehman-collapsed.jpg
Let's retry the GET request now with the updated string.

In [129]:
get_request_url = 'https://paper-api.alpaca.markets/v2/puzzle/202108/hint/' + str(final_msg_string)
print("GET request URL: " + get_request_url)

#* Perfrom GET request on this URL:
headers = {'APCA-API-KEY-ID': config.API_KEY, 'APCA-API-SECRET-KEY': config.API_SECRET}
r = requests.get(get_request_url, headers=headers)
r_json = r.json()
print("Request JSON response: " + str(r_json))

GET request URL: https://paper-api.alpaca.markets/v2/puzzle/202108/hint/110-days-after-lehman-collapsed.jpg
Request JSON response: {'url': 'https://s3.amazonaws.com/prod.alpaca.markets/puzzle/202108/110-days-after-lehman-collapsed.jpg?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAI6TTUPCZPCYJXPJA%2F20210908%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210908T000943Z&X-Amz-Expires=900&X-Amz-SignedHeaders=host&X-Amz-Signature=80bffac2299f5c4075b4a2d4bbf7abe30886c4468a30f7e8a0297164e1155fcc'}


It worked! And the GET request returned this URL, which links us to an image of the front page of The Times No 69523 on January 3, 2009.
<br><img src="screenshots/GET_Request_Return.png" alt="GET Request Return" title="GET Request Return" /><br>

OK... but what now? The URL seems to suggest that there are some headers, let's see what's in them:

In [130]:
returned_headers = r.headers
print("Request response headers: " + str(returned_headers))

Request response headers: {'Server': 'nginx', 'Date': 'Wed, 08 Sep 2021 00:09:43 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Length': '395', 'Connection': 'keep-alive', 'Access-Control-Allow-Origin': '*', 'Vary': 'Origin', 'X-Alpaca-Puzzle': "h = this.GetHash(); x = {last 10-digit prime found in consecutive digits of h in base-10}; requests.post('https://api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': account.id.int * x, 'email': your_waitlist_email})", 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'}


Ohhh! Looks like there is a ```X-Alpaca-Puzzle``` key, let's get that and take a look.

In [131]:
puzzle_clue = returned_headers['X-Alpaca-Puzzle'];
print(puzzle_clue)

puzzle_clue2 = 'X-Alpaca-Puzzle: ' + puzzle_clue
print(puzzle_clue2)

h = this.GetHash(); x = {last 10-digit prime found in consecutive digits of h in base-10}; requests.post('https://api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': account.id.int * x, 'email': your_waitlist_email})
X-Alpaca-Puzzle: h = this.GetHash(); x = {last 10-digit prime found in consecutive digits of h in base-10}; requests.post('https://api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': account.id.int * x, 'email': your_waitlist_email})


## 7) Understanding the Headers
OK, it looks like we are on the home stretch. We seem to have our final instructions:
1. ```h = this.GetHash()```
2. ```x = {last 10-digit prime found in consecutive digits of h in base-10}```
3. ```requests.post('https://api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': account.id.int * x, 'email': your_waitlist_email})```

### 7.1) h = this.GetHash()
This is honestly the trickiest part of this whole challenge. It took me the longest to figure out, but I eventually got there.

The first step in the right direction was just googling 'The Times No 69523'. The first thing that search returned was another Times article stating how that issue is effectively a collectors item now:
<br><img src="screenshots/TheTimesNo69523.png" alt="TheTimesNo69523" title="The Times No 69523" /><br>

So, we see that the first bitcoins had the headline of The Times No 69523 as part of the original block hash.

The next step was to google more about this original block hash, which led me to the wiki on the Genesis block:
<br><img src="screenshots/GenesisGetHash.png" alt="Genesis Block" title="Genesis Block Wiki" /><br>

And there we see it! The `GetHash()` we have to be looking for!
<br><img src="screenshots/GenesisHashHeadline.png" alt="Genesis Hash" title="Genesis Hash" /><br>

Now let's take that and set it equal to h (where h = genesis_hash below):

In [132]:
genesis_hash_hex = '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f'
print("The Genesis block hash: " + str(genesis_hash_hex))
genesis_hash = int(genesis_hash_hex, 16)
print("Request base-10 hash: " + str(genesis_hash))

The Genesis block hash: 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f
Request base-10 hash: 10628944869218562084050143519444549580389464591454674019345556079


### 7.2) x = {last 10-digit prime found in consecutive digits of h in base-10}
OK, this is a bit more complicated, but still fairly straightforward.

First, let's define the following function:
- Function: ```get_primes_list(number, digits)```
- Inputs:
    - ```number```
        - An integer that we will search for primes by looking at consecutive digits of the number itself.
    - ```digits```
        - An integer that specifies the number of digits to search along ```number``` for prime numbers.
- Output:
    - ```prime_list```
        - A list of prime numbers found in consecutive digits on ```number```

In [133]:
from math import sqrt

def get_primes_list(number, digits):
    #* Cast 'number' into string for indexing
    num_string = str(number)
    num_string_len = len(num_string)
    
    #* Check to see that num_string_len is less than digits, or else this doesn't make sense
    if num_string_len < digits:
        return ['number does not have enough digits to proceed!']
    
    #* Initiate empty prime list
    prime_list = []
    
    #* Index holder
    ind = 0
    
    while ind < (num_string_len - (digits - 1)):
        #* Check if substrings are prime.
        #* We will check from 2 to sqrt(n). 
        #* If we find any factor then it is not a prime number.
        
        prime_flag = 0 #* This flag maintains status whether the substring is prime or not
        
        cum_num = int(num_string[ind:ind+digits]) #* Number that is a substring of 'number'
        
        if(cum_num > 1): #* We should only look if number is 2 or greater
            for i in range(2, int(sqrt(cum_num)) + 1):
                if (cum_num % i == 0):
                    prime_flag = 1 #* Not prime
                    break
            if (prime_flag == 0): #* Prime
                print("Prime found! Added to list!")
                prime_list.append(cum_num)
                ind += 1
            else:
                ind += 1
        else:
            print("Not > 1!")
    return prime_list

primes = get_primes_list(genesis_hash, 10)
print("Our 'get_primes_list' function output: " + str(primes))

Prime found! Added to list!
Prime found! Added to list!
Our 'get_primes_list' function output: [4050143519, 9345556079]


Now we just need to get the last prime added to the list ```primes```:

In [134]:
last_prime = primes[-1]
print("Last prime: " + str(last_prime))

Last prime: 9345556079


### 7.3 Answer
The answer is calculated as ```account.id.int * last_prime```. 
#### 7.3.1 Get account.id
Let's do another GET request to the Alpaca API endpoint for account info.

In [135]:
get_request_account_url = 'https://paper-api.alpaca.markets/v2/account'
print("GET account request URL: " + get_request_account_url)

#* Perfrom GET request on this URL:
r = requests.get(get_request_account_url, headers=headers)
r_json = r.json()

#* Retrieve account ID
accnt_id_hex_raw = r_json['id']
print("Hex Account ID Raw: " + accnt_id_hex_raw)

GET account request URL: https://paper-api.alpaca.markets/v2/account
Hex Account ID Raw: 781fa3ed-2841-404a-899e-b1a98f9255c4


#### 7.3.2) Convert account.id to Integer

In [136]:
#* Strip ID of dashes to get it into what seems to be hexidecimal
accnt_id_hex = accnt_id_hex_raw.replace('-','')
print("Hex Account ID Cleaned: " + accnt_id_hex)

#* Convert to base-10 integer
accnt_id = int(accnt_id_hex, 16)
print("Account ID Integer: " + str(accnt_id))

Hex Account ID Cleaned: 781fa3ed2841404a899eb1a98f9255c4
Account ID Integer: 159671645519102679685427683220203394500


#### 7.3.3) Calculate Final Answer
We can now calculate the answer: ```accnt_id * last_prime```.

In [137]:
answer = accnt_id * last_prime
print("The final answer is: " + str(answer))

The final answer is: 1492220317424983158759338492633458129085910165500


## 8) Post the Answer per Instruction
The last step is to post the answer per: ```requests.post('https://api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': account.id.int * x, 'email': your_waitlist_email})```

In [138]:
solution = requests.post('https://paper-api.alpaca.markets/v2/puzzle/202108/solve', json={'answer': answer, 'email': 'roderpad@gmail.com'}, headers=headers)
print("Request JSON response: " + str(solution.json()))

Request JSON response: {'message': 'Congratulations! Your answer will be verified and points will be awarded to your account.'}


## 9) Celebrate! We Did It!
Our post returns the following response: `{'message': 'Congratulations! Your answer will be verified and points will be awarded to your account.'}`
<iframe src="https://giphy.com/embed/kHOwxTcKHbPDK6LZ2k" width="480" height="480" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/easy-peasy-lemon-squeezy-neveen-ms-kHOwxTcKHbPDK6LZ2k"></a></a></p>